File: backend/mcp-code-intelligence-kotlin/src/main/kotlin/com/codeintel/analyzers/complexity/counters/KotlinCounter.kt
/**
 * KSA-161: Kotlin AST node counter.
 */
package com.codeintel.analyzers.complexity.counters

import com.codeintel.analyzers.complexity.BaseNodeCounter

class KotlinCounter : BaseNodeCounter() {
    override val language = "kotlin"
    override val branchNodeTypes = setOf(
        "if_expression", "when_entry", "elvis_expression",
    )
    override val loopNodeTypes = setOf(
        "for_statement", "while_statement", "do_while_statement",
    )
    override val logicalOperators = setOf("&&", "||")
    override val exceptionNodeTypes = setOf("catch_block")

    override fun getReturnNodeTypes(): Set<String> =
        setOf("jump_expression") // Kotlin uses jump_expression for return
}

File: backend/src/engine/analyzers/complexity/counters/KotlinCounter.ts
/**
 * KSA-161: Kotlin decision point counter.
 */

import type { SyntaxNode } from '../../../parsers/types.js';
import { BaseNodeCounter } from './BaseNodeCounter.js';

export class KotlinCounter extends BaseNodeCounter {
  readonly language = 'kotlin';
  readonly branchNodeTypes = [
    'if_expression',
    'when_entry',
    'elvis_expression',
  ];
  readonly loopNodeTypes = [
    'for_statement',
    'while_statement',
    'do_while_statement',
  ];
  readonly logicalOperators = ['&&', '||'];
  readonly exceptionNodeTypes = ['catch_block'];

  protected isLogicalOp(node: SyntaxNode): boolean {
    if (node.type === 'conjunction_expression' || node.type === 'disjunction_expression') {
      return true;
    }
    if (node.type === 'binary_expression') {
      const op = node.childForFieldName('operator');
      if (op && this.logicalOperators.includes(op.text)) return true;
    }
    return false;
  }

  protected getReturnNodeTypes(): string[] {
    return ['jump_expression']; // Kotlin uses jump_expression for return
  }
}
